src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs

```diff
@@ -590,6 +590,211 @@ public void TestEventNotificationOrder()
+            }).Dispose();
+        }
+
+        [ConditionalFact(typeof(RemoteExecutor), nameof(RemoteExecutor.IsSupported))]
+        public void TestSamplerTagsWithMixedListenerModels()
+        {
+            RemoteExecutor.Invoke(() => {
+                const string w3cId = "00-99d43cb30a4cdb4fbeee3a19c29201b0-e82825765f051b47-01";
+
+                using ActivitySource aSource = new ActivitySource("SamplerTags6");
+                using ActivityListener listener1 = new ActivityListener();
+                using ActivityListener listener2 = new ActivityListener();
+                listener1.ShouldListenTo = (activitySource) => activitySource.Name == "SamplerTags6";
+                listener2.ShouldListenTo = (activitySource) => activitySource.Name == "SamplerTags6";
+
+                // listener1 provide SampleUsingParentId callback and doesn't provide Sample
+                listener1.SampleUsingParentId = (ref ActivityCreationOptions<string> activityOptions) =>
+                {
+                    activityOptions.SamplingTags.Add("tag1", "value1");
+                    return ActivitySamplingResult.AllData;
+                };
+
+                // listener2 provide Sample callback and doesn't provide SampleUsingParentId
+                // Sample should get called as we convert the w3c id to a context.
+                listener2.Sample = (ref ActivityCreationOptions<ActivityContext> activityOptions) =>
+                {
+                    activityOptions.SamplingTags.Add("tag2", "value2");
+                    return ActivitySamplingResult.AllData;
+                };
+
+                ActivitySource.AddActivityListener(listener1);
+                ActivitySource.AddActivityListener(listener2);
+
+                using Activity activity = aSource.StartActivity("a1", ActivityKind.Client, w3cId);
+
+                Assert.NotNull(activity);
+                Assert.Equal(2, activity.TagObjects.Count());
+                Assert.Equal(new KeyValuePair<string, object>("tag1", "value1"), activity.TagObjects.ElementAt(0));
+                Assert.Equal(new KeyValuePair<string, object>("tag2", "value2"), activity.TagObjects.ElementAt(1));
+            }).Dispose();
+        }
+
+        [ConditionalFact(typeof(RemoteExecutor), nameof(RemoteExecutor.IsSupported))]
+        public void TestAddSamplerAndActivityCreationTags()
+        {
+            RemoteExecutor.Invoke(() => {
+                using ActivitySource aSource = new ActivitySource("SamplerTags7");
+                using ActivityListener listener = new ActivityListener();
+                listener.ShouldListenTo = (activitySource) => activitySource.Name == "SamplerTags7";
+
+                listener.Sample = (ref ActivityCreationOptions<ActivityContext> activityOptions) =>
+                {
+                    activityOptions.SamplingTags.Add("SamplerTag1", "SamplerValue1");
```

Comment:
If I set an "tag1" to null in here, will it result in the removal of "tag1" which was provided at the time of startActivity?

Reply:
No, we don't remove any tags from the original set. All tags provided by the sampler will get added regardless the values of teh tags. It is hard to know the intention of the sampler want to add tags with null value or want to remove old tags in the original set.